esp/services/ws_dfu/ws_dfuService.hpp

```diff
@@ -30,6 +30,81 @@
 #include "fvrelate.hpp"
 #include "dadfs.hpp"
 
+class CThorNodeGroup: public CInterface
+{
+    CDateTime timeCached;
+    StringAttr groupName;
+    unsigned keyhash;
+    unsigned nodeCount;
+    bool replicateOutputs;
+
+public:
+    IMPLEMENT_IINTERFACE;
+    CThorNodeGroup(const char* _groupName, unsigned _nodeCount, bool _replicateOutputs)
+        : groupName(_groupName), nodeCount(_nodeCount), replicateOutputs(_replicateOutputs)
+    {
+        keyhash = hashc((const byte *)groupName.get(),groupName.length(),0);
+        timeCached.setNow();
+    }
+
+    inline unsigned queryHash() { return keyhash; }
+    inline const char *queryGroupName() { return groupName.get(); }
+
+    inline bool queryNodeCount() { return nodeCount; };
+    inline bool queryReplicateOutputs() { return replicateOutputs; };
+    inline bool queryCanReplicate() { return replicateOutputs && (nodeCount > 1); };
+    inline bool checkTimeOut(unsigned timeOutMinutes)
+    {
+        CDateTime timeLine;
+        timeLine.setNow();
+        timeLine.adjustTime(-timeOutMinutes);
+        return timeCached <= timeLine;
+    }
+};
+
+class CThorNodeGroupCache: public SuperHashTableOf<CThorNodeGroup, const char>
+{
+    CriticalSection sect;
+    CThorNodeGroup* readNodeGroup(const char* _groupName);
+
+public:
+    IMPLEMENT_IINTERFACE;
+
+    ~CThorNodeGroupCache() { releaseAll(); }
+
+    inline void onAdd(void *e)
+    {
+        // not used
+    }
+
+    inline void onRemove(void *e)
+    {
+        // not used
```

Comment:
Have you checked that the old elements are correctly released when they are replaced?